src/edit.rs

```diff
@@ -1,28 +1,69 @@
 use encoding::types::EncodingRef;
 
+use std::u32;
 use error::Error;
 use history::{Cursor, History};
 use buffer::Buffer;
 use parser::{parse, ParseError, ParseSuccess};
 use instr;
 
+#[derive(Copy,Clone)]
+pub enum EditMode {
+    Emacs,
+    Vi,
+}
+
+#[derive(Copy, Clone, PartialEq)]
+pub enum ViMode {
+    Insert,
+    Normal,
+    Replace,
+    MoveChar(instr::CharMoveType),
+    DeleteMoveChar(instr::CharMoveType),
+    ChangeMoveChar(instr::CharMoveType),
+    Delete,
+    Change,
+}
+
+/// Set a new vi mode based on the current vi mode.
+///
+/// There are several common mode transitions that we handle here.
+macro_rules! set_next_vi_mode {
+    ( $ctx:ident ) => {
+        $ctx.vi_mode = match $ctx.vi_mode {
+            ViMode::Delete => ViMode::Normal,
+            ViMode::Change => ViMode::Insert,
+            ViMode::DeleteMoveChar(_) => ViMode::Normal,
+            ViMode::ChangeMoveChar(_) => ViMode::Insert,
+            _ => ViMode::Normal,
+        }
+    }
+}
+
 pub struct EditCtx<'a> {
     buf: Buffer,
     history_cursor: Cursor<'a>,
     prompt: &'a str,
     seq: Vec<u8>,
-    enc: EncodingRef
+    enc: EncodingRef,
+    mode: EditMode,
+    vi_mode: ViMode,
+    vi_count: u32,
```

Comment:
I did two versions of this. One with EditModeState (which I think I'll rename to ModeState) with a struct as follows:

    enum EditModeState {
        Emacs,
        Vi { mode: ViMode, count: u32 },
    }

And one with a tuple type as follows:

    enum EditModeState {
        Emacs,
        Vi(ViMode, u32),
    }

The second version (with the tuple) is cleaner than the first, as rust allows matching directly with the inner tuple, where as the struct version requires you to decompose before matching the inner values. But both require more lines of code than the version with separate vars. I am going to look at it some more to see if I can find a way to simplify things some more.